exercises/Toluwalase_Agoro/ex01_basics/src/main.cpp

```diff
@@ -0,0 +1,137 @@
+#include <iostream>
+#include <vector>
+#include <fstream>
+#include <string>
+#include <iomanip>
+#include <algorithm>
+
+
+
+std::string getWord(std::ifstream& stream) {
+  /* Function accept an input stream object and return a string from
+  input stream by reading stream 1 character at a time  and formats it  
+  by making it lowercase and removing punctuation marks then 
+  concatenates it to a string until a space ,newline or '-'  character.
+  */
+  std::string outputString{};
+  char inputCharacter{};
+
+  while (true) {
+    stream.get(inputCharacter);
+    if (inputCharacter == '-' || inputCharacter == ' ' || inputCharacter == '\n' || stream.eof())
+      break;
+    if (inputCharacter == '"' || inputCharacter == '!' || inputCharacter == '('
+      || inputCharacter == ')' || inputCharacter == ':' || inputCharacter == '\''
+      || inputCharacter == ',' || inputCharacter == '?' || inputCharacter == '.')
+      continue;
+
+    outputString += tolower(inputCharacter);
+  }
+  return outputString;
+}
+
+int main() {
+  /*This block of code declares all vectors used
+  and gets input from user so a to open
+  input file
+  */
+  std::vector <std::string> vectorOfWords{};
+  std::vector <int> vectorOfCounts{};
+  std::vector <int> vectorOfOrderToRead{};
+  char filePathToInput[100];
+  std::cout << "Please input a file path to input file :";
+  std::cin.getline(filePathToInput, 100);
+  std::ifstream fileToRead;
+  fileToRead.open(filePathToInput);
+
+  if (!fileToRead) {
+    /* This if block checks if file was opened correctly
+    and if it was not program is terminated
+    */
+    std::cout << "File could not be opened" << std::endl;
+    std::cout << "Ensure that file path is valid" << std::endl;
+  }
+  while (!fileToRead.eof()) {
+    /* This while block reads input file and adds formatted words to
+    VectorOfWords until the end of file has been reached. This block
+    has an if , else if block that determines whether word is more than
+    4 letters and already exists in VectorOfWords. if it does it increments
+    the count in VectorOfCounts otherwise it is added to VectorOfWords
+    */
+    std::string holderString{ };
+    holderString = getWord(fileToRead);
+    auto location = std::find(vectorOfWords.begin(), vectorOfWords.end(), holderString);
+    if (holderString.size() > 4 && location == vectorOfWords.end()) {
+      vectorOfWords.push_back(holderString);
+      vectorOfCounts.push_back(1);
+    }
+    else {
+      if (holderString.size() > 4 && location != vectorOfWords.end())
+      {
+        vectorOfCounts[std::distance(vectorOfWords.begin(), location)] += 1;
+      }
+    }
+  }
+  fileToRead.close();
+  int maxCount{ 0 };
+  /*Next Block of code implements a variation of counting
+  sort  on vectorOfCounts to produce filled vectorOfOrderToRead*/
+  for (auto& count : vectorOfCounts) {
```

Comment:
Thanks for making the change here - now much faster.

When most people do the sorting at this stage of the exercise they use the std::sort function in the standard library. I can see you've used a counting sort method here instead and I may be persuaded on this point - please could you let me know why this might be better than using the std::sort function?

Even if we stick with your version it would be good practice to make it easy to swap in the std::sort function at some point in the future. Plus this would also fulfil part of the point of the exercise which is to demonstrate you know how to call the std::sort function. So with this in mind please could you do these changes:

a) pull the counting sort code out into a separate function called countingSort or similar
b) the std::sort function would require a single vector containing the words\counts. So please could you replace these two vectors with a vector of pairs:
```  
  std::vector <std::string> vectorOfWords{};
  std::vector <int> vectorOfCounts{};
```
The "pairs" vector would be:
```
  std::vector <std::pair<std::string,int>> vectorOfWords{};
```
I think your counting sort code can be adapted to work with this arrangement

c) please can you write a second sorting function called stdSort or similar that uses the std::sort function. This function can be left unused once I've checked on your reasoning re counting sort